### ISSUE TITLE
Tensorflow GetStarted - 1

### ISSUE THREAD

**Issue by killelder** [killelder/tensorflownote#1]
https://www.tensorflow.org/get_started/get_started

------------------------------------------------------------
TensorFlow是一個編程系統, 是由兩個部分結合的, 構造和執行
計算圖是由一系列的operations作用在node之間所組成的

ex.1-1
node1 = tf.constant(3.0, dtype=tf.float32)
node2 = tf.constant(4.0) # also tf.float32 implicitly
print(node1, node2)

>> Tensor("Const:0", shape=(), dtype=float32) Tensor("Const_1:0", shape=(), dtype=float32)

node並沒有吐出3,4,
由此可見node還沒有經過計算,
要開始計算, 必須要建立一個Session
這樣才會開始執行計算圖中的運算

ex.1-2
sess = tf.Session()
print(sess.run([node1, node2]))

>> [3.0, 4.0]

ex.1-3
node3 = tf.add(node1, node2)
print("node3:", node3)
print("sess.run(node3):", sess.run(node3))

>> node3: Tensor("Add:0", shape=(), dtype=float32)
>> sess.run(node3): 7.0

在範例1-3中可以更清楚的看出這一點
"建立"計算圖, 與"執行"計算圖的兩個部分

placeholder可以先建立好, 之後再餵值,

ex. 2-1
a = tf.placeholder(tf.float32)
b = tf.placeholder(tf.float32)
adder_node = a + b  # + provides a shortcut for tf.add(a, b)
print(sess.run(adder_node, {a: 3, b: 4.5}))
print(sess.run(adder_node, {a: [1, 3], b: [2, 4]}))

>> 7.5
>> [ 3.  7.]

為了可以訓練, 我們需要改變graph以至於獲得不同的output結果, 
Variables讓我們可以添加可以訓練的參數進到graph裡,

ex. 3-1
W = tf.Variable([.3], dtype=tf.float32)
b = tf.Variable([-.3], dtype=tf.float32)
x = tf.placeholder(tf.float32)
linear_model = W*x + b

tf.constant會在呼叫的時候被初始化, 並且不能變更裡面的值,
tf.Variables 則不會在呼叫的時候進行初始化, 必須呼叫global_variables_initializer, 並且在sess.run()
的時候才會初始化

ex. 3-2
init = tf.global_variables_initializer()
sess.run(init)
print(sess.run(linear_model, {x: [1, 2, 3, 4]}))

>> [ 0.          0.30000001  0.60000002  0.90000004]

為了要建立一個model, 我們還需要一個y當答案, 還有一個loss function來評分

ex. 3-3
y = tf.placeholder(tf.float32)
squared_deltas = tf.square(linear_model - y)
loss = tf.reduce_sum(squared_deltas)
print(sess.run(loss, {x: [1, 2, 3, 4], y: [0, -1, -2, -3]}))

>> 23.66

tf.square 對tensor做平方
tf.reduce_sum 對tenosr做平均

我們可以自己修改W,b的值,
利用assign
ex.3-4
fixW = tf.assign(W, [-1.])
fixb = tf.assign(b, [1.])
sess.run([fixW, fixb])
print(sess.run(loss, {x: [1, 2, 3, 4], y: [0, -1, -2, -3]}))
